# Лабораторная работа №3: Использование библиотеки символьных вычислений SymPy для решения задач

Выполнил студент группы 9381 Колованов Родион Алексеевич, Вариант 7.

## Цель работы

Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

## Задачи работы

1. Используя методы библиотеки SymPy решить диофантово уравнение;
2. Используя методы библиотеки SymPy по открытой части ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово;
3. Используя методы библиотеки SymPy записать в ДНФ и КНФ булеву функцию, заданную аналитическим выражением.

## Выполнение работы

Для начала подключим нужные библиотеки:

In [18]:
import sympy as sp
import sympy.crypto.crypto as spc
import sympy.logic.boolalg as spl

### Решение диофантова уравнения

Задано следующее диофантово уравнение: $1742x+1911y=65$. Решим его при помощи метода *diophantine* библиотеки SymPy:

In [19]:
x, y = sp.symbols("x y")
sp.diophantine(1742 * x + 1911 * y - 65)

{(147*t_0 - 170, 155 - 134*t_0)}

Получили следующий ответ:

$x=-170+147t_0$

$y=155-134t_0$

$t_0 \in \mathbb{Z}$

### Расшифровка закодированного сообщения

Дан открытый ключ $(e = 5, m = 39)$ и закодированное сообщение $(28, 10, 32, 11)$. Найдем закрытый ключ при помощи метода *rsa_private_key* и расшифруем сообщение при помощи метода *decipher_rsa* библиотеки SymPy:

In [20]:
alphabet = {2: "А", 3: "Б", 4: "В", 5: "Г", 6: "Д", 7: "Е", 8: "Ж", 9: "З", 10: "И", 11: "Й", 12: "К", 13: "Л", 14: "М", 15: "Н", 16: "О", 17: "П", 18: "Р", 19: "С", 20: "Т", 21: "У", 22: "Ф", 23: "Х", 24: "Ц", 25: "Ч", 26: "Ш", 27: "Щ", 28: "Ы", 29: "Ь", 30: "Э", 31: "Ю", 32: "Я"}
encoded_message = [28, 10, 32, 11]
private_key = spc.rsa_private_key(39, 5)

for character in encoded_message:
    print(alphabet[spc.decipher_rsa(character, private_key)], end='')   

СВАТ

Получили следующий ответ: *СВАТ*

### Преобразование булевой функции в ДНФ И КНФ

Дана следующая булева функция: $f(x,y,z)=((x\ xor\ z)(yx)) \vee (y\ xor\ (z \vee y))$. Найдем ДНФ и КНФ при помощи методов *to_dnf* и *to_cnf* библиотеки SymPy:

In [21]:
x, y, z = sp.symbols("x y z")

print("ДНФ: {}".format(spl.to_dnf(((x ^ z) & (y & x)) | (y ^ (z | y)), simplify=True)))
print("КНФ: {}".format(spl.to_cnf(((x ^ z) & (y & x)) | (y ^ (z | y)), simplify=True)))

ДНФ: (z & ~y) | (x & y & ~z)
КНФ: (x | z) & (y | z) & (~y | ~z)


Получили следующий ответ:

ДНФ: $f(x,y,z)=(z \wedge \bar y) \vee (x \wedge y \wedge \bar z)$

КНФ: $f(x,y,z)=(x \vee z) \wedge (y \vee z) \wedge (\bar y \vee \bar z)$

## Выводы

Были получены навыки использования библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики. В ходе лабораторной работы были решены следующие задачи:

* Решение диофантова уравнения;
* Расшифровка закодированного сообщения;
* Запись булевой функции, заданную аналитическим выражением, в виде ДНФ и КНФ.